File: go/sigil/redaction.go
package sigil

import (
	"regexp"
	"strings"
)

// GenerationSanitizer mutates a generation before export. Sanitizers receive
// the fully normalized Generation and return the version to ship. Implementations
// may mutate strings/payloads (e.g. redact secrets) but must preserve message
// and part structure.
//
// If a sanitizer panics during Recorder.End the SDK downgrades the generation's
// content capture mode to ContentCaptureModeMetadataOnly and logs a warning.
type GenerationSanitizer func(Generation) Generation

// SecretRedactionOptions configures the built-in secret-redaction sanitizer.
type SecretRedactionOptions struct {
	// RedactInputMessages, when true, also sanitizes user messages in
	// Generation.Input. Assistant and tool messages in Input are sanitized
	// regardless because they typically replay tool results and prior model
	// output that share the same secret surface as Generation.Output.
	RedactInputMessages bool
	// RedactEmailAddresses, when non-nil, sets whether email addresses are
	// redacted. Nil defaults to true (redact). Set to a pointer to false to
	// preserve email addresses verbatim.
	RedactEmailAddresses *bool
}

type secretPattern struct {
	id string
	re *regexp.Regexp
}

// tier1Patterns covers high-confidence secret formats. Applied by both the
// full and lightweight redactors.
var tier1Patterns = []secretPattern{
	{"grafana-cloud-token", regexp.MustCompile(`\bglc_[A-Za-z0-9_-]{20,}`)},
	{"grafana-service-account-token", regexp.MustCompile(`\bglsa_[A-Za-z0-9_-]{20,}`)},
	{"aws-access-token", regexp.MustCompile(`\b(?:A3T[A-Z0-9]|AKIA|ASIA|ABIA|ACCA)[A-Z2-7]{16}\b`)},
	{"github-pat", regexp.MustCompile(`\bghp_[A-Za-z0-9_]{36,}`)},
	{"github-oauth", regexp.MustCompile(`\bgho_[A-Za-z0-9_]{36,}`)},
	{"github-app-token", regexp.MustCompile(`\bghs_[A-Za-z0-9_]{36,}`)},
	{"github-fine-grained-pat", regexp.MustCompile(`\bgithub_pat_[A-Za-z0-9_]{82}`)},
	{"anthropic-api-key", regexp.MustCompile(`\bsk-ant-api03-[a-zA-Z0-9_-]{93}AA`)},
	{"anthropic-admin-key", regexp.MustCompile(`\bsk-ant-admin01-[a-zA-Z0-9_-]{93}AA`)},
	{"openai-api-key", regexp.MustCompile(`\bsk-[a-zA-Z0-9]{20}T3BlbkFJ[a-zA-Z0-9]{20}`)},
	{"openai-project-key", regexp.MustCompile(`\bsk-proj-[a-zA-Z0-9_-]{40,}`)},
	{"openai-svcacct-key", regexp.MustCompile(`\bsk-svcacct-[a-zA-Z0-9_-]{40,}`)},
	{"gcp-api-key", regexp.MustCompile(`\bAIza[A-Za-z0-9_-]{35}`)},
	{"private-key", regexp.MustCompile(`-----BEGIN[A-Z ]*PRIVATE KEY-----[\s\S]*?-----END[A-Z ]*PRIVATE KEY-----`)},
	{"connection-string", regexp.MustCompile(`(?:postgres|mysql|mongodb|redis|amqp)://[^\s'"]+@[^\s'"]+`)},
	{"bearer-token", regexp.MustCompile(`[Bb]earer\s+[A-Za-z0-9_.\-~+/]{20,}={0,3}`)},
	{"slack-token", regexp.MustCompile(`\bxox[bporas]-[A-Za-z0-9-]{10,}`)},
	{"stripe-key", regexp.MustCompile(`\b[sr]k_(?:live|test)_[A-Za-z0-9]{20,}`)},
	{"sendgrid-api-key", regexp.MustCompile(`\bSG\.[A-Za-z0-9_-]{22}\.[A-Za-z0-9_-]{43}`)},
	{"twilio-api-key", regexp.MustCompile(`\bSK[a-f0-9]{32}`)},
	{"npm-token", regexp.MustCompile(`\bnpm_[A-Za-z0-9]{36}`)},
	{"pypi-token", regexp.MustCompile(`\bpypi-[A-Za-z0-9_-]{50,}`)},
}

// tier1Combined alternates all tier1Patterns into a single regex so each input
// is scanned once instead of once per pattern. Each pattern is wrapped in a
// capturing group; the matched group index identifies which pattern fired.
var tier1Combined = func() *regexp.Regexp {
	parts := make([]string, len(tier1Patterns))
	for i, p := range tier1Patterns {
		parts[i] = "(" + p.re.String() + ")"
	}
	return regexp.MustCompile(strings.Join(parts, "|"))
}()

// emailPattern is toggleable via SecretRedactionOptions.RedactEmailAddresses.
var emailPattern = secretPattern{
	id: "email",
	re: regexp.MustCompile(`(?i)\b[A-Z0-9._%+\-]+@[A-Z0-9.\-]+\.[A-Z]{2,}\b`),
}

// envSecretPattern covers heuristic env-style key/value assignments. Applied
// only by the full redactor (skipped for assistant text/thinking in
// lightweight mode) to avoid mangling natural-language strings.
var envSecretPattern = secretPattern{
	id: "env-secret-value",
	re: regexp.MustCompile(`(?i)((?:PASSWORD|SECRET|TOKEN|KEY|CREDENTIAL|API_KEY|PRIVATE_KEY|ACCESS_KEY)\s*[=:]\s*)([^\s"{}\[\],]+)`),
}

// redactFull applies tier 1, optional email, and env-style patterns. Use for
// tool-call args, tool-result content, system prompts, and any field where
// arbitrary content can be expected (env dumps, shell output).
func redactFull(s string, includeEmail bool) string {
	s = redactTier1String(s)
	if includeEmail {
		s = emailPattern.re.ReplaceAllString(s, "[REDACTED:"+emailPattern.id+"]")
	}
	return envSecretPattern.re.ReplaceAllString(s, "${1}[REDACTED:"+envSecretPattern.id+"]")
}

// redactLight applies tier 1 and optional email patterns only. Use for
// assistant text and reasoning, where env-style heuristics would cause too many
// false positives, and for short metadata strings (titles, error messages).
func redactLight(s string, includeEmail bool) string {
	s = redactTier1String(s)
	if includeEmail {
		s = emailPattern.re.ReplaceAllString(s, "[REDACTED:"+emailPattern.id+"]")
	}
	return s
}

// redactFullBytes is the []byte form of redactFull; it operates on the source
// slice directly so JSON payloads avoid a string round-trip.
func redactFullBytes(src []byte, includeEmail bool) []byte {
	src = redactTier1Bytes(src)
	if includeEmail {
		src = emailPattern.re.ReplaceAll(src, []byte("[REDACTED:"+emailPattern.id+"]"))
	}
	return envSecretPattern.re.ReplaceAll(src, []byte("${1}[REDACTED:"+envSecretPattern.id+"]"))
}

func redactTier1String(s string) string {
	matches := tier1Combined.FindAllStringSubmatchIndex(s, -1)
	if len(matches) == 0 {
		return s
	}
	var b strings.Builder
	b.Grow(len(s))
	last := 0
	for _, m := range matches {
		b.WriteString(s[last:m[0]])
		for g := 1; g <= len(tier1Patterns); g++ {
			if m[2*g] >= 0 {
				b.WriteString("[REDACTED:")
				b.WriteString(tier1Patterns[g-1].id)
				b.WriteByte(']')
				break
			}
		}
		last = m[1]
	}
	b.WriteString(s[last:])
	return b.String()
}

func redactTier1Bytes(src []byte) []byte {
	matches := tier1Combined.FindAllSubmatchIndex(src, -1)
	if len(matches) == 0 {
		return src
	}
	out := make([]byte, 0, len(src))
	last := 0
	for _, m := range matches {
		out = append(out, src[last:m[0]]...)
		for g := 1; g <= len(tier1Patterns); g++ {
			if m[2*g] >= 0 {
				out = append(out, "[REDACTED:"...)
				out = append(out, tier1Patterns[g-1].id...)
				out = append(out, ']')
				break
			}
		}
		last = m[1]
	}
	out = append(out, src[last:]...)
	return out
}

// NewSecretRedactionSanitizer returns a GenerationSanitizer that redacts
// known secret formats from generation content. The returned sanitizer is
// safe for concurrent use.
//
// By default it redacts Generation.Output (assistant + tool), Generation.SystemPrompt,
// Generation.ConversationTitle, Generation.CallError, and the assistant /
// tool messages in Generation.Input. User messages in Generation.Input are
// only redacted when RedactInputMessages is set. Email redaction is on unless
// RedactEmailAddresses points to false.
func NewSecretRedactionSanitizer(opts SecretRedactionOptions) GenerationSanitizer {
	includeEmail := opts.RedactEmailAddresses == nil || *opts.RedactEmailAddresses
	redactInputs := opts.RedactInputMessages

	return func(g Generation) Generation {
		if g.SystemPrompt != "" {
			g.SystemPrompt = redactFull(g.SystemPrompt, includeEmail)
		}
		// ConversationTitle and CallError are short natural-language strings;
		// lightweight redaction (tier 1 + email) avoids mangling them with the
		// env-style heuristic.
		if g.ConversationTitle != "" {
			g.ConversationTitle = redactLight(g.ConversationTitle, includeEmail)
		}
		if g.CallError != "" {
			g.CallError = redactLight(g.CallError, includeEmail)
		}

		for i := range g.Input {
			sanitizeMessage(&g.Input[i], inputTextMode(g.Input[i].Role, redactInputs), includeEmail)
		}

		for i := range g.Output {
			sanitizeMessage(&g.Output[i], outputTextMode(g.Output[i].Role), includeEmail)
		}

		return g
	}
}

// textMode is which tier to apply to PartKindText for a given role; thinking,
// tool-call, and tool-result parts use a fixed tier regardless.
type textMode int

const (
	textModeSkip textMode = iota
	textModeLight
	textModeFull
)

// inputTextMode picks the tier for an Input message's text part. Historic
// assistant turns and tool results in Input always get role-aware redaction;
// user text is only redacted when the caller opts in.
func inputTextMode(role Role, redactUserInput bool) textMode {
	switch role {
	case RoleUser:
		if redactUserInput {
			return textModeFull
		}
		return textModeSkip
	case RoleTool:
		return textModeFull
	case RoleAssistant:
		return textModeLight
	default:
		return textModeSkip
	}
}

func outputTextMode(role Role) textMode {
	switch role {
	case RoleAssistant:
		return textModeLight
	case RoleTool:
		return textModeFull
	default:
		return textModeSkip
	}
}

func sanitizeMessage(m *Message, mode textMode, includeEmail bool) {
	if mode == textModeSkip {
		return
	}
	for i := range m.Parts {
		p := &m.Parts[i]
		switch p.Kind {
		case PartKindText:
			if mode == textModeFull {
				p.Text = redactFull(p.Text, includeEmail)
			} else {
				p.Text = redactLight(p.Text, includeEmail)
			}
		case PartKindThinking:
			p.Thinking = redactLight(p.Thinking, includeEmail)
		case PartKindToolCall:
			if p.ToolCall != nil && len(p.ToolCall.InputJSON) > 0 {
				p.ToolCall.InputJSON = redactFullBytes(p.ToolCall.InputJSON, includeEmail)
			}
		case PartKindToolResult:
			if p.ToolResult != nil {
				if p.ToolResult.Content != "" {
					p.ToolResult.Content = redactFull(p.ToolResult.Content, includeEmail)
				}
				if len(p.ToolResult.ContentJSON) > 0 {
					p.ToolResult.ContentJSON = redactFullBytes(p.ToolResult.ContentJSON, includeEmail)
				}
			}
		}
	}
}

File: dotnet/src/Grafana.Agento11y/Redaction.cs
using System.Text;
using System.Text.RegularExpressions;

namespace Grafana.Sigil;

/// <summary>
/// Mutates a normalized generation before export. Sanitizers may redact strings
/// and byte payloads, but should preserve message and part structure.
/// </summary>
public delegate Generation GenerationSanitizer(Generation generation);

/// <summary>Options for the built-in secret redaction sanitizer.</summary>
public sealed class SecretRedactionOptions
{
    /// <summary>
    /// Redact user messages in <see cref="Generation.Input"/>. <c>null</c>
    /// falls back to <c>AGENTO11Y_REDACT_INPUT_MESSAGES</c> (legacy
    /// <c>SIGIL_REDACT_INPUT_MESSAGES</c>), then <c>false</c>.
    /// Assistant and tool messages in input are always sanitized.
    /// </summary>
    public bool? RedactInputMessages { get; set; }

    /// <summary>
    /// Redact generic email addresses. Defaults to <c>true</c>.
    /// </summary>
    public bool RedactEmailAddresses { get; set; } = true;
}

/// <summary>Factory for the built-in regex-based secrets redactor.</summary>
public static class SecretRedactionSanitizer
{
    private static readonly EnvPair EnvRedactInputMessages = new(
        "AGENTO11Y_REDACT_INPUT_MESSAGES",
        "SIGIL_REDACT_INPUT_MESSAGES"
    );
    private static readonly HashSet<string> TrueTokens = new(StringComparer.OrdinalIgnoreCase)
    {
        "1", "true", "yes", "on",
    };
    private static readonly HashSet<string> FalseTokens = new(StringComparer.OrdinalIgnoreCase)
    {
        "0", "false", "no", "off",
    };

    private sealed record SecretPattern(string Id, Regex Regex);

    private static readonly SecretPattern[] Tier1Patterns =
    [
        new("grafana-cloud-token", new Regex(@"\bglc_[A-Za-z0-9_-]{20,}", RegexOptions.Compiled)),
        new("grafana-service-account-token", new Regex(@"\bglsa_[A-Za-z0-9_-]{20,}", RegexOptions.Compiled)),
        new("aws-access-token", new Regex(@"\b(?:A3T[A-Z0-9]|AKIA|ASIA|ABIA|ACCA)[A-Z2-7]{16}\b", RegexOptions.Compiled)),
        new("github-pat", new Regex(@"\bghp_[A-Za-z0-9_]{36,}", RegexOptions.Compiled)),
        new("github-oauth", new Regex(@"\bgho_[A-Za-z0-9_]{36,}", RegexOptions.Compiled)),
        new("github-app-token", new Regex(@"\bghs_[A-Za-z0-9_]{36,}", RegexOptions.Compiled)),
        new("github-fine-grained-pat", new Regex(@"\bgithub_pat_[A-Za-z0-9_]{82}", RegexOptions.Compiled)),
        new("anthropic-api-key", new Regex(@"\bsk-ant-api03-[a-zA-Z0-9_-]{93}AA", RegexOptions.Compiled)),
        new("anthropic-admin-key", new Regex(@"\bsk-ant-admin01-[a-zA-Z0-9_-]{93}AA", RegexOptions.Compiled)),
        new("openai-api-key", new Regex(@"\bsk-[a-zA-Z0-9]{20}T3BlbkFJ[a-zA-Z0-9]{20}", RegexOptions.Compiled)),
        new("openai-project-key", new Regex(@"\bsk-proj-[a-zA-Z0-9_-]{40,}", RegexOptions.Compiled)),
        new("openai-svcacct-key", new Regex(@"\bsk-svcacct-[a-zA-Z0-9_-]{40,}", RegexOptions.Compiled)),
        new("gcp-api-key", new Regex(@"\bAIza[A-Za-z0-9_-]{35}", RegexOptions.Compiled)),
        new("private-key", new Regex(@"-----BEGIN[A-Z ]*PRIVATE KEY-----[\s\S]*?-----END[A-Z ]*PRIVATE KEY-----", RegexOptions.Compiled)),
        new("connection-string", new Regex(@"(?:postgres|mysql|mongodb|redis|amqp)://[^\s'""]+@[^\s'""]+", RegexOptions.Compiled)),
        new("bearer-token", new Regex(@"[Bb]earer\s+[A-Za-z0-9_.\-~+/]{20,}={0,3}", RegexOptions.Compiled)),
        new("slack-token", new Regex(@"\bxox[bporas]-[A-Za-z0-9-]{10,}", RegexOptions.Compiled)),
        new("stripe-key", new Regex(@"\b[sr]k_(?:live|test)_[A-Za-z0-9]{20,}", RegexOptions.Compiled)),
        new("sendgrid-api-key", new Regex(@"\bSG\.[A-Za-z0-9_-]{22}\.[A-Za-z0-9_-]{43}", RegexOptions.Compiled)),
        new("twilio-api-key", new Regex(@"\bSK[a-f0-9]{32}", RegexOptions.Compiled)),
        new("npm-token", new Regex(@"\bnpm_[A-Za-z0-9]{36}", RegexOptions.Compiled)),
        new("pypi-token", new Regex(@"\bpypi-[A-Za-z0-9_-]{50,}", RegexOptions.Compiled)),
    ];

    private static readonly SecretPattern EmailPattern = new(
        "email",
        new Regex(@"\b[A-Z0-9._%+\-]+@[A-Z0-9.\-]+\.[A-Z]{2,}\b", RegexOptions.Compiled | RegexOptions.IgnoreCase)
    );

    private static readonly SecretPattern EnvSecretPattern = new(
        "env-secret-value",
        new Regex(
            @"((?:PASSWORD|SECRET|TOKEN|KEY|CREDENTIAL|API_KEY|PRIVATE_KEY|ACCESS_KEY)\s*[=:]\s*)([^\s""{}\[\],]+)",
            RegexOptions.Compiled | RegexOptions.IgnoreCase
        )
    );

    /// <summary>
    /// Returns a reusable generation sanitizer that redacts known secret formats.
    /// </summary>
    public static GenerationSanitizer Create(SecretRedactionOptions? options = null)
    {
        return Create(options, Environment.GetEnvironmentVariable, null);
    }

    internal static GenerationSanitizer Create(
        SecretRedactionOptions? options,
        Func<string, string?> envLookup,
        Action<string>? logger
    )
    {
        var resolved = options ?? new SecretRedactionOptions();
        var redactInputs = ResolveRedactInputMessages(resolved.RedactInputMessages, envLookup, logger);
        var includeEmail = resolved.RedactEmailAddresses;

        return generation =>
        {
            if (!string.IsNullOrEmpty(generation.SystemPrompt))
            {
                generation.SystemPrompt = RedactFull(generation.SystemPrompt, includeEmail);
            }

            if (!string.IsNullOrEmpty(generation.ConversationTitle))
            {
                generation.ConversationTitle = RedactLight(generation.ConversationTitle, includeEmail);
            }

            if (!string.IsNullOrEmpty(generation.CallError))
            {
                generation.CallError = RedactLight(generation.CallError, includeEmail);
            }

            foreach (var message in generation.Input)
            {
                SanitizeMessage(message, InputTextMode(message.Role, redactInputs), includeEmail);
            }

            foreach (var message in generation.Output)
            {
                SanitizeMessage(message, OutputTextMode(message.Role), includeEmail);
            }

            return generation;
        };
    }

    internal static string RedactFull(string value, bool includeEmail)
    {
        var result = RedactTier1(value);
        if (includeEmail)
        {
            result = ApplyPattern(result, EmailPattern);
        }

        return EnvSecretPattern.Regex.Replace(result, $"$1[REDACTED:{EnvSecretPattern.Id}]");
    }

    internal static string RedactLight(string value, bool includeEmail)
    {
        var result = RedactTier1(value);
        if (includeEmail)
        {
            result = ApplyPattern(result, EmailPattern);
        }

        return result;
    }

    private static byte[] RedactFullBytes(byte[] value, bool includeEmail)
    {
        if (value.Length == 0)
        {
            return value;
        }

        return Encoding.UTF8.GetBytes(RedactFull(Encoding.UTF8.GetString(value), includeEmail));
    }

    private static string RedactTier1(string value)
    {
        var result = value;
        foreach (var pattern in Tier1Patterns)
        {
            result = ApplyPattern(result, pattern);
        }

        return result;
    }

    private static string ApplyPattern(string value, SecretPattern pattern)
    {
        return pattern.Regex.Replace(value, $"[REDACTED:{pattern.Id}]");
    }

    private static void SanitizeMessage(Message message, TextMode mode, bool includeEmail)
    {
        if (mode == TextMode.Skip)
        {
            return;
        }

        foreach (var part in message.Parts)
        {
            switch (part.Kind)
            {
                case PartKind.Text:
                    part.Text = mode == TextMode.Full
                        ? RedactFull(part.Text, includeEmail)
                        : RedactLight(part.Text, includeEmail);
                    break;
                case PartKind.Thinking:
                    part.Thinking = RedactLight(part.Thinking, includeEmail);
                    break;
                case PartKind.ToolCall:
                    if (part.ToolCall?.InputJson is { Length: > 0 })
                    {
                        part.ToolCall.InputJson = RedactFullBytes(part.ToolCall.InputJson, includeEmail);
                    }
                    break;
                case PartKind.ToolResult:
                    if (part.ToolResult != null)
                    {
                        if (!string.IsNullOrEmpty(part.ToolResult.Content))
                        {
                            part.ToolResult.Content = RedactFull(part.ToolResult.Content, includeEmail);
                        }
                        if (part.ToolResult.ContentJson.Length > 0)
                        {
                            part.ToolResult.ContentJson = RedactFullBytes(part.ToolResult.ContentJson, includeEmail);
                        }
                    }
                    break;
            }
        }
    }

    private static TextMode InputTextMode(MessageRole role, bool redactUserInput)
    {
        return role switch
        {
            MessageRole.User => redactUserInput ? TextMode.Full : TextMode.Skip,
            MessageRole.Assistant => TextMode.Light,
            MessageRole.Tool => TextMode.Full,
            _ => TextMode.Skip,
        };
    }

    private static TextMode OutputTextMode(MessageRole role)
    {
        return role switch
        {
            MessageRole.Assistant => TextMode.Light,
            MessageRole.Tool => TextMode.Full,
            _ => TextMode.Skip,
        };
    }

    private static bool ResolveRedactInputMessages(
        bool? explicitValue,
        Func<string, string?> envLookup,
        Action<string>? logger
    )
    {
        if (explicitValue.HasValue)
        {
            return explicitValue.Value;
        }

        var value = EnvConfig.EnvTrimmed(envLookup, EnvRedactInputMessages, out var key);
        if (value == null)
        {
            return false;
        }

        if (TrueTokens.Contains(value))
        {
            return true;
        }

        if (FalseTokens.Contains(value))
        {
            return false;
        }

        logger?.Invoke($"sigil: ignoring invalid {key}: {value}");
        return false;
    }

    private enum TextMode
    {
        Skip,
        Light,
        Full,
    }
}
